crates/ra_syntax/src/ast/generated.rs

```diff
@@ -9,4648 +9,3387 @@
+    type Repr = rowan::SyntaxNode<RaTypes>;
 }
 
-impl<'a> AstNode<'a> for EnumVariant<'a> {
-    fn cast(syntax: SyntaxNodeRef<'a>) -> Option<Self> {
+impl AstNode for EnumVariant {
+    fn cast(syntax: &SyntaxNode) -> Option<&Self> {
         match syntax.kind() {
-            ENUM_VARIANT => Some(EnumVariant { syntax }),
+            ENUM_VARIANT => Some(EnumVariant::from_repr(syntax.into_repr())),
             _ => None,
         }
     }
-    fn syntax(self) -> SyntaxNodeRef<'a> { self.syntax }
-}
-
-impl<R: TreeRoot<RaTypes>> EnumVariantNode<R> {
-    pub fn borrowed(&self) -> EnumVariant {
-        EnumVariantNode { syntax: self.syntax.borrowed() }
-    }
-    pub fn owned(&self) -> EnumVariantNode {
-        EnumVariantNode { syntax: self.syntax.owned() }
-    }
+    fn syntax(&self) -> &SyntaxNode { &self.syntax }
+    fn to_owned(&self) -> TreePtr<EnumVariant> { TreePtr::cast(self.syntax.to_owned()) }
 }
 
 
-impl<'a> ast::NameOwner<'a> for EnumVariant<'a> {}
-impl<'a> EnumVariant<'a> {
-    pub fn expr(self) -> Option<Expr<'a>> {
+impl ast::NameOwner for EnumVariant {}
+impl EnumVariant {
+    pub fn expr(&self) -> Option<&Expr> {
         super::child_opt(self)
     }
 }
 
 // EnumVariantList
-#[derive(Debug, Clone, Copy,)]
-pub struct EnumVariantListNode<R: TreeRoot<RaTypes> = OwnedRoot> {
-    pub(crate) syntax: SyntaxNode<R>,
-}
-pub type EnumVariantList<'a> = EnumVariantListNode<RefRoot<'a>>;
-
-impl<R1: TreeRoot<RaTypes>, R2: TreeRoot<RaTypes>> PartialEq<EnumVariantListNode<R1>> for EnumVariantListNode<R2> {
-    fn eq(&self, other: &EnumVariantListNode<R1>) -> bool { self.syntax == other.syntax }
+#[derive(Debug, PartialEq, Eq, Hash)]
+#[repr(transparent)]
+pub struct EnumVariantList {
+    pub(crate) syntax: SyntaxNode,
 }
-impl<R: TreeRoot<RaTypes>> Eq for EnumVariantListNode<R> {}
-impl<R: TreeRoot<RaTypes>> Hash for EnumVariantListNode<R> {
-    fn hash<H: Hasher>(&self, state: &mut H) { self.syntax.hash(state) }
+unsafe impl TransparentNewType for EnumVariantList {
+    type Repr = rowan::SyntaxNode<RaTypes>;
 }
 
-impl<'a> AstNode<'a> for EnumVariantList<'a> {
-    fn cast(syntax: SyntaxNodeRef<'a>) -> Option<Self> {
+impl AstNode for EnumVariantList {
+    fn cast(syntax: &SyntaxNode) -> Option<&Self> {
         match syntax.kind() {
-            ENUM_VARIANT_LIST => Some(EnumVariantList { syntax }),
+            ENUM_VARIANT_LIST => Some(EnumVariantList::from_repr(syntax.into_repr())),
             _ => None,
         }
     }
-    fn syntax(self) -> SyntaxNodeRef<'a> { self.syntax }
-}
-
-impl<R: TreeRoot<RaTypes>> EnumVariantListNode<R> {
-    pub fn borrowed(&self) -> EnumVariantList {
-        EnumVariantListNode { syntax: self.syntax.borrowed() }
-    }
-    pub fn owned(&self) -> EnumVariantListNode {
-        EnumVariantListNode { syntax: self.syntax.owned() }
-    }
+    fn syntax(&self) -> &SyntaxNode { &self.syntax }
+    fn to_owned(&self) -> TreePtr<EnumVariantList> { TreePtr::cast(self.syntax.to_owned()) }
```

Comment:
and those `AstNode` implementations too with proc_macro.

Reply:
These all are already generated from a `grammar.ron` with a script. I feel that a good old code generation would work better than a proc macro here:

* easy to understand, bc you see the generated code
* fast to compile (you don't generate the impls on every compilation)
* easy to modify (template is just a tera file)

Reply:
oops, my fault, didn't notice the `generated.rs` name.